mtcpclient/groupofconnections.go

```diff
@@ -46,6 +55,60 @@ func (gc GroupOfConnections) AtLeastOneConnectionInError() bool {
 	return gc.isIn(tcpclient.ConnectionError)
 }
 
-func (gc GroupOfConnections) AtLeastOneConnectionEstablished() bool {
-	return gc.isIn(tcpclient.ConnectionEstablished)
+func (gc GroupOfConnections) AtLeastOneConnectionOK() bool {
+	return gc.isIn(tcpclient.ConnectionEstablished) || gc.isIn(tcpclient.ConnectionClosed)
+}
+
+func appendConnections(previous_connections []tcpclient.Connection, new_connections ...tcpclient.Connection) []tcpclient.Connection {
+	// TODO implement as a method using as a reference
+	m := len(previous_connections)
+	n := m + len(new_connections)
+	if n > cap(previous_connections) {
+		// allocate double what's needed, for future growth.
+		newSlice := make([]tcpclient.Connection, (n+1)*2)
+		copy(newSlice, previous_connections)
+		previous_connections = newSlice
+	}
+	previous_connections = previous_connections[0:n]
+	copy(previous_connections[m:n], new_connections)
+	return previous_connections
+}
+
+func (gc GroupOfConnections) getFilteredListByStatus(statuses []tcpclient.ConnectionStatus) (filteredConnections []tcpclient.Connection) {
+	for _, connection := range gc.connections {
+		for _, s := range statuses {
+			if connection.GetConnectionStatus() == s {
+				filteredConnections = appendConnections(filteredConnections, connection)
+			}
+		}
+	}
+	return filteredConnections
+}
+
+func (gc GroupOfConnections) pingStyleReport() (output string) {
+	var introduction string
+	var filteredConnections GroupOfConnections
+	var mr metricsCollectionStats
+
+	if gc.AtLeastOneConnectionOK() {
+		filteredConnections.connections = gc.getFilteredListByStatus([]tcpclient.ConnectionStatus{tcpclient.ConnectionEstablished, tcpclient.ConnectionClosed})
+		mr = filteredConnections.calculateMetricsReport()
+		introduction = "Response time stats for " + strconv.Itoa(mr.numberOfConnections) +
+			" successful connections min/avg/max/dev = "
+	}
+
+	if gc.AtLeastOneConnectionInError() {
+		filteredConnections.connections = gc.getFilteredListByStatus([]tcpclient.ConnectionStatus{tcpclient.ConnectionError})
+		mr = filteredConnections.calculateMetricsReport()
+		introduction = "Time to error stats for " + strconv.Itoa(mr.numberOfConnections) +
+			" failed connections min/avg/max/dev = "
+	}
+
+	output = introduction +
```

Comment:
I think there is a problem with this implementation: a group of connections can have one connection ok AND one connection in error... and you need to generate something like two reports in that case, right? Run the tests I wrote in the parallel branch: they may detect this condition.
In any case, I'd suggest to move the responsability to the caller and make this more specific:
* I take the "introduction" keywords as an argument
* I generate the intro with the keywords and the actual numbers
* If the caller wants one report for a filtered list of connectionsThatWentWell and another for connectionsThatFailed, its responsibility of the caller to do this logic